src/output-json-modbus.c

```diff
@@ -0,0 +1,386 @@
+/* Copyright (C) 2017 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+/**
+ * \file
+ *
+ * \author Eric Leblond <eric@regit.org>
+ *
+ * Implements Modbus JSON logging portion of the engine.
+ *
+ */
+
+#include "suricata-common.h"
+#include "debug.h"
+#include "detect.h"
+#include "pkt-var.h"
+#include "conf.h"
+
+#include "threads.h"
+#include "threadvars.h"
+#include "tm-threads.h"
+
+#include "util-unittest.h"
+#include "util-buffer.h"
+#include "util-debug.h"
+#include "util-byte.h"
+
+#include "output.h"
+#include "output-json.h"
+#include "output-json-modbus.h"
+
+#include "app-layer.h"
+#include "app-layer-parser.h"
+
+#include "app-layer-modbus.h"
+
+#ifdef HAVE_LIBJANSSON
+#include <jansson.h>
+
+#define LOG_MODBUS_NO_FLAG     0
+#define LOG_MODBUS_USE_DICT    (1<<0)
+
+typedef struct LogModbusFileCtx_ {
+    LogFileCtx *file_ctx;
+    uint32_t    flags;
+} LogModbusFileCtx;
+
+typedef struct LogModbusLogThread_ {
+    LogModbusFileCtx *modbuslog_ctx;
+    uint32_t            count;
+    MemBuffer          *buffer;
+} LogModbusLogThread;
+
+static json_t *JsonModbusLogRegisters(ModbusTransaction *modbustx,
+                                      uint8_t quantity, uint8_t use_dict)
+{
+    uint16_t val;
+    int i = 0;
+    if (use_dict) {
+        char key[6];
+        json_t *data_dict = json_object();
+        if (data_dict == NULL)
+            goto end;
+        if (unlikely(modbustx->data == NULL))
```

Comment:
if this fails data_dict will be lost

Reply:
it seems all paths check modbustx->data before getting here, so the check can probably be removed